File: ios/Pets/Domain/Sources/Domain/List/Reducers/ListReducer.swift
//
//  File.swift
//  
//
//  Created by Dylan Velez on 10/08/21.
//

import Foundation
import Architecture
import Combine
import Services

public struct ListReducer: Reducing, Depending {
    public typealias State = ListState
    public typealias Event = ListEvent
    
    public struct Dependencies {
        public var listService: DataService
        public var favoriteService: FavoriteService
        
        public init(listService: DataService,
                    favoriteService: FavoriteService) {
            self.listService = listService
            self.favoriteService = favoriteService
        }
    }
    
    public var dependencies: Dependencies
    
    public init(listService: DataService, favoriteService: FavoriteService) {
        self.dependencies = Dependencies(listService: listService, favoriteService: favoriteService)
    }
    
    public func reduce(state: inout State, forEvent event: Event) -> Effect<Event> {
        switch event {
        case .fetchList:
            state.list = []
            state.isLoading = true
        case .fetchListCompleted(let newPets):
            state.list = newPets
            state.isLoading = false
        case .fetchFavorites:
            state.favorites = []
        case .fetchFavoritesCompleted(let favorites):
            state.favorites =  favorites
        case .addToFavorites(let id):
            if let favorite = state.list.first(where: {$0.id == id}) {
                state.favorites.append(favorite)
            }
        case .removeFromFavorites(let id):
            state.favorites.removeAll { $0.id == id }
        }
        
        return sideEffect(event: event)
    }
    
    private func sideEffect(event: Event) -> Effect<Event> {
        switch event {
        case .fetchList:
            return dependencies.listService.getList()
                .map(Event.fetchListCompleted)
                .replaceError(with: .fetchListCompleted(pets: []))
                .eraseToAnyPublisher()
        case .fetchFavorites:
            return dependencies.favoriteService.fetchList()
                .map(Event.fetchFavoritesCompleted)
                .replaceError(with: Event.fetchFavoritesCompleted(pets: []))
                .eraseToAnyPublisher()
        case .addToFavorites(let id):
            dependencies.favoriteService.addToFavorites(id: id)
        case .removeFromFavorites(let id):
            dependencies.favoriteService.removeFromFavorites(id: id)
        case .fetchListCompleted, .fetchFavoritesCompleted:
            return nil
        }
        return nil
    }
}

File: android/Pets/app/src/main/java/com/pets/domain/list/ListReducer.kt
package com.pets.domain.list

import androidx.lifecycle.MutableLiveData
import com.pets.architecture.common.Depending
import com.pets.architecture.reducers.Reducing
import com.pets.models.Pet
import com.pets.services.ServiceLocator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.withContext

class ListReducer: Reducing<ListState, ListEvent>, Depending<ListReducer.Dependencies> {
    class Dependencies {
        val listService = ServiceLocator.dataService
        val favoriteService = ServiceLocator.favoriteService
    }

    override val dependencies: Dependencies
        get() = ListReducer.Dependencies()

    override suspend fun reduce(state: MutableLiveData<ListState>, event: ListEvent): ListEvent? =
        withContext(Dispatchers.Main) {
            val emptyList = emptyList<Pet>().toMutableList()
            when (event) {
                is FetchList -> {
                    state.postValue(
                        ListState(
                            list = emptyList,
                            isLoading = true,
                            favorites = state.value?.favorites ?: emptyList
                        )
                    )
                }

                is FetchListCompleted -> {
                    state.postValue(
                        ListState(
                            list = event.list.toMutableList(),
                            isLoading = false,
                            favorites = state.value?.favorites ?: emptyList
                        )
                    )
                }

                is FetchFavorites -> {
                    state.postValue(
                        ListState(
                            list = state.value?.list ?: emptyList,
                            isLoading = state.value?.isLoading ?: false,
                            favorites = emptyList
                        )
                    )
                }

                is FetchFavoriteCompleted -> {
                    state.postValue(
                        ListState(
                            list = state.value?.list ?: emptyList,
                            isLoading = state.value?.isLoading ?: false,
                            favorites = event.list.toMutableList()
                        )
                    )
                }

                is AddToFavorites -> {
                    val favorite = state.value?.list?.firstOrNull { it.id == event.id }
                    favorite?.let {
                        val newFavorites = state.value?.favorites ?: emptyList
                        newFavorites.add(it)
                        state.postValue(
                            ListState(
                                list = state.value?.list ?: emptyList,
                                isLoading = state.value?.isLoading ?: false,
                                favorites = newFavorites
                            )
                        )
                    }
                }

                is RemoveFromFavorites -> {
                    val favorites = state.value?.favorites ?: emptyList
                    favorites.removeAll { it.id == event.id }
                    state.postValue(
                        ListState(
                            list = state.value?.list ?: emptyList,
                            isLoading = state.value?.isLoading ?: false,
                            favorites = favorites
                        )
                    )
                }
            }

            return@withContext sideEffect(event = event)
        }

    private suspend fun sideEffect(event: ListEvent): ListEvent? {
        when (event) {
            is FetchList -> {
                val list = dependencies.listService.getList()
                return FetchListCompleted(list = list)
            }

            is FetchFavorites -> {
                val list = dependencies.favoriteService.fetchList()
                return FetchFavoriteCompleted(list = list)
            }

            is AddToFavorites -> {
                dependencies.favoriteService.addToFavorites(id = event.id)
            }

            is RemoveFromFavorites -> {
                dependencies.favoriteService.removeFromFavorites(id = event.id)
            }

            else -> return null
        }
        return  null
    }
}